src/simulate-swap.ts

```diff
@@ -0,0 +1,927 @@
+
+  while (chunk > limitingChunk || (chunk === limitingChunk && bit >= limitingBit)) {
+    const value = tickmap.get(chunk) ?? 0n
+
+    const upper = (value >> HALF_CHUNK_SIZE) & 0xffffffffffffffffffffffffffffffffn
+    const lower = value & 0xffffffffffffffffffffffffffffffffn
+
+    let part = 0n
+    let part_bit = 0n
+
+    if (bit >= HALF_CHUNK_SIZE) {
+      part = upper
+      part_bit = bit - HALF_CHUNK_SIZE
+    } else {
+      part = lower
+      part_bit = bit
+    }
+
+    let mask = 1n << part_bit
+    while (part_bit > 0n && !(part & mask)) {
+      mask = mask >> 1n
+      part_bit = part_bit - 1n
+    }
+
+    if (part & mask) {
+      if (bit >= HALF_CHUNK_SIZE) {
+        bit = part_bit + HALF_CHUNK_SIZE
+      } else {
+        bit = part_bit
+      }
+      if (chunk > limitingChunk || (chunk == limitingChunk && bit >= limitingBit)) {
+        const index = chunk * CHUNK_SIZE + bit
+        return [true, (index - GLOBAL_MAX_TICK / tickSpacing) * tickSpacing]
+      } else {
+        return [false, 0n]
+      }
+    }
+
+    // check lower part
+    if (bit == HALF_CHUNK_SIZE) {
+      part = lower
+      part_bit = HALF_CHUNK_SIZE - 1n
+      mask = 1n << part_bit
+
+      while (part_bit > 0n && !(part & mask)) {
+        mask = mask >> 1n
+        part_bit -= 1n
+      }
+
+      if (part & mask) {
+        bit = part_bit
+        if (chunk > limitingChunk || (chunk == limitingChunk && bit >= limitingBit)) {
+          let index = chunk * CHUNK_SIZE + bit
+          return [true, (index - GLOBAL_MAX_TICK / tickSpacing) * tickSpacing]
+        } else {
+          return [false, 0n]
+        }
+      }
+    }
+
+    // move to next chunk
+    if (chunk > 0n) {
+      chunk -= 1n
+    } else {
+      return [false, 0n]
+    }
+    bit = CHUNK_SIZE - 1n
+  }
+  return [false, 0n]
+}
+
+const tickToPosition = (tick: bigint, tickSpacing: bigint): [bigint, bigint] => {
+  if (tick < GLOBAL_MIN_TICK || tick > GLOBAL_MAX_TICK) {
+    throw new Error(String(InvariantError.InvalidTickIndex))
+  }
+
+  const bitmapIndex = (tick + GLOBAL_MAX_TICK) / tickSpacing
+  const chunk = bitmapIndex / CHUNK_SIZE
+  const bit = bitmapIndex % CHUNK_SIZE
+
+  return [chunk, bit]
+}
+
+const getSearchLimit = (tick: bigint, tickSpacing: bigint, up: boolean): bigint => {
+  const index = tick / tickSpacing
+
+  let limit = 0n
+  if (up) {
+    const rangeLimitUp = index + SEARCH_RANGE
+    const sqrtPriceLimitUp = GLOBAL_MAX_TICK / tickSpacing
+
+    if (rangeLimitUp < sqrtPriceLimitUp) {
+      limit = rangeLimitUp
+    } else {
+      limit = sqrtPriceLimitUp
+    }
+  } else {
+    const rangeLimitDown = index - SEARCH_RANGE
+    const sqrtPriceLimitDown = GLOBAL_MIN_TICK / tickSpacing
+
+    if (rangeLimitDown > sqrtPriceLimitDown) {
+      limit = rangeLimitDown
+    } else {
+      limit = sqrtPriceLimitDown
+    }
+  }
+  return limit * tickSpacing
+}
+
+const computeSwapStep = (
```

Comment:
Same as comments above, there is few places in this function where ternary operator can be used

Reply:
I will add some, but only if I deem it simplifies instead of complicating the function. In some places where a ternary operator could be used after a few modifications I am also very much against adding new variables.